src/v/storage/log_manager.cc

```diff
@@ -292,13 +292,15 @@ ss::future<> log_manager::dispatch_topic_dir_deletion(ss::sstring dir) {
                 if (!exists) {
                     return ss::now();
                 }
-                return directory_walker::empty(std::filesystem::path(dir))
-                  .then([dir](bool empty) {
-                      if (!empty) {
-                          return ss::now();
-                      }
-                      return ss::remove_file(dir);
-                  });
+                return ss::sync_directory(dir).then([dir] {
+                    return directory_walker::empty(std::filesystem::path(dir))
+                      .then([dir](bool empty) {
+                          if (!empty) {
+                              return ss::now();
+                          }
+                          return ss::remove_file(dir);
```

Comment:
I could imagine that we want to sync the parent directory after removing `dir` so that it doesn't reappear after a crash. But the motivation for syncing the directory before removing it is unclear. The commit message makes it sound like syncing the directory will make removed partition "visible" is some way, but prior to a crash the file system should remain consistent even if the directory isn't synced.

am I missing something? maybe the problem/solution description needs to be more specific?

Reply:
what's the perf impact here. could we get around by writing another file `.deleted` with some content in it. so we don' force a directory sync (parent dir is expensive, child dir is ok)

Reply:
the problem that i see is that `directory_walker::empty(std::filesystem::path(dir))` reports the directory as empty, but then in `ss::remove_file` an exception is thrown. In general nothing bad happens as we will simply retry removal but exception is logged and causing test to fail with `BadLog`

Reply:
> the problem that i see is that directory_walker::empty(std::filesystem::path(dir)) reports the directory as empty, but then in ss::remove_file an exception is thrown. In general nothing bad happens as we will simply retry removal but exception is logged and causing test to fail with BadLog

I would expect that the file system metadata to be consistent without requiring fsync (e.g. the same way that you can observe a file size increasing after writing without an fsync).

One thing to notice is that `empty` and `remove_file` don't run atomically together. It is conceivable there is a race here. Should we add some logging / debugging either into the ducktape test or into the log manager that would help us root cause this issue?

I suppose it is possible there is some real issue that isn't merely a race that fsync fixes, but my hunch would be that that is something wrong in seastar list_directory and not with posix.